File: mac/ClickBridgeMac/AppState.swift
import Foundation
import AppKit

struct PairingActionPresentation: Equatable {
    let title: String
    let requiresReplacementConfirmation: Bool

    init(status: PairStatus?) {
        let replacing = status?.requiresReplacementConfirmation == true
        title = replacing ? "Replace Phone" : "Pair Phone"
        requiresReplacementConfirmation = replacing
    }
}

@MainActor
final class AppState: ObservableObject {
    @Published private(set) var connection: RelayClient.Status = .disconnected
    @Published private(set) var permission: PermissionState = .unknown
    @Published private(set) var lastResult = "—"
    @Published var notice: String?
    @Published private(set) var pairing: PairingController?
    @Published private(set) var phoneManagement: PhoneManagementController?
    @Published private(set) var pairingAction = PairingActionPresentation(status: nil)
    @Published private(set) var relaySetupRetryAvailable = false

    let settings: SettingsStore
    private let client: RelayClient
    private let processor: ActionProcessor
    private let permissionService: PostEventPermissionService
    private let activationNotifications: NotificationCenter
    private let enroll: @Sendable () async throws -> DesktopEnrollment
    private var activationObserver: NSObjectProtocol?
    private var credentialRevision: UInt64 = 0
    private var credentialTask: Task<Void, Never>?
    private var credentialEligible = true
    private var automaticEnrollmentAttempted = false
    private var lastStatusSequence: UInt64 = 0

    init(settings: SettingsStore, client: RelayClient, processor: ActionProcessor,
         permissionService: PostEventPermissionService,
         activationNotifications: NotificationCenter = .default,
         initialNotice: String? = nil,
         enroll: @escaping @Sendable () async throws -> DesktopEnrollment = {
             try await SessionEnrollmentService.enroll()
         }) {
        self.settings = settings
        self.client = client
        self.processor = processor
        self.permissionService = permissionService
        self.activationNotifications = activationNotifications
        self.enroll = enroll
        notice = initialNotice
        permission = permissionService.isGranted() ? .ready : .required
        activationObserver = activationNotifications.addObserver(
            forName: NSApplication.didBecomeActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            Task { @MainActor in self?.refreshPermission() }
        }
        let initialCredentialRevision = beginCredentialOperation()
        credentialTask = Task { [weak self, client, processor, settings] in
            await client.setStatusHandler { [weak self] event in
                Task { @MainActor in
                    await self?.handleStatusEvent(event)
                }
            }
            await client.setResultHandler { [weak self] result in
                Task { @MainActor in
                    self?.lastResult = "\(result.status.rawValue): \(result.reason.rawValue)"
                    self?.refreshPermission()
                }
            }
            await client.setPairingHandler { [weak self] message in
                await self?.receivePairing(message)
            }
            await client.setPhoneManagementHandler { [weak self] message in
                await self?.phoneManagement?.receive(message)
            }
            await processor.setRemoteEnabled(settings.remoteEnabled)
            guard let self, initialCredentialRevision == self.credentialRevision,
                  !Task.isCancelled else { return }
            await self.publishState()
            guard initialCredentialRevision == self.credentialRevision, !Task.isCancelled else { return }
            self.connect(credentialRevision: initialCredentialRevision, allowAutomaticEnrollment: true)
        }
    }

    deinit {
        credentialTask?.cancel()
        if let activationObserver { activationNotifications.removeObserver(activationObserver) }
    }

    var remoteToggleEnabled: Bool { true }

    @discardableResult
    func reconnect() -> Task<Void, Never> {
        let revision = beginCredentialOperation()
        return connect(credentialRevision: revision)
    }

    @discardableResult
    func retryRelaySetup() -> Task<Void, Never> {
        let revision = beginCredentialOperation()
        credentialEligible = true
        relaySetupRetryAvailable = false
        return enrollAndConnect(credentialRevision: revision, automaticRecoveryUsed: true)
    }

    func publishState() async {
        await client.updateAdvertisedState(MacState(remoteEnabled: settings.remoteEnabled, permission: permission))
    }

    func setPersistedRemoteEnabled(_ enabled: Bool) {
        Task {
            await processor.setRemoteEnabled(enabled)
            settings.remoteEnabled = enabled
            await publishState()
        }
    }

    func refreshPermission() {
        permission = permissionService.isGranted() ? .ready : .required
        Task { await publishState() }
    }

    func requestPermission() {
        _ = permissionService.requestFromUserAction()
        refreshPermission()
    }

    func beginPairing() {
        guard let pairing else { return }
        Task { await pairing.beginPairing() }
    }

    func confirmReplacement() {
        guard let pairing else { return }
        Task {
            await pairing.beginPairing()
            await pairing.confirmReplacement()
        }
    }

    @discardableResult
    func saveToken(_ token: String) -> Task<Void, Never> {
        let revision = beginCredentialOperation()
        do {
            try settings.saveMacToken(token)
            credentialEligible = true
            notice = "Relay enrollment saved."
            return connect(credentialRevision: revision)
        }
        catch {
            credentialEligible = false
            notice = settings.storageError
            let task = Task { _ = await client.clearConfigurationAndStop(credentialRevision: revision) }
            credentialTask = task
            return task
        }
    }

    @discardableResult
    func clearToken() -> Task<Void, Never> {
        let revision = beginCredentialOperation()
        credentialEligible = false
        var persistenceError: String?
        do { try settings.clearMacToken() }
        catch { persistenceError = settings.storageError }

        let task = Task {
            guard revision == credentialRevision, !Task.isCancelled else { return }
            let cleared = await client.clearConfigurationAndStop(credentialRevision: revision)
            guard cleared else { return }
            guard revision == credentialRevision, !Task.isCancelled else { return }
            notice = persistenceError ?? "Relay enrollment cleared."
        }
        credentialTask = task
        return task
    }

    private func beginCredentialOperation() -> UInt64 {
        credentialRevision &+= 1
        return credentialRevision
    }

    func handleStatusEvent(_ event: RelayClient.StatusEvent) async {
        guard event.credentialRevision == credentialRevision,
              event.sequence > lastStatusSequence else { return }
        lastStatusSequence = event.sequence
        connection = event.status
        if event.cause == .authenticationRejected {
            recoverAfterAuthenticationRejection()
            return
        }
        if event.status == .connected {
            relaySetupRetryAvailable = false
            do {
                try settings.clearAutomaticRecoveryMarker()
            } catch {
                notice = settings.storageError
            }
        }
        if let pairing {
            await pairing.refreshStatus(capabilityAvailable: event.status == .connected)
        }
        guard event.credentialRevision == credentialRevision,
              event.sequence == lastStatusSequence else { return }
    }

    private func receivePairing(_ message: WireMessage) async {
        guard let pairing else { return }
        let previousState = pairing.state
        await pairing.receive(message)
        if case .pairStatus(let status) = message,
           previousState == .checkingStatus, pairing.state == .ready {
            pairingAction = PairingActionPresentation(status: status)
        }
    }

    @discardableResult
    private func connect(
        credentialRevision revision: UInt64,
        allowAutomaticEnrollment: Bool = false
    ) -> Task<Void, Never> {
        let task = Task {
            do { try Task.checkCancellation() }
            catch { return }
            guard revision == credentialRevision else { return }
            guard credentialEligible else {
                await client.clearConfigurationAndStop(credentialRevision: revision)
                return
            }
            let record: DesktopEnrollmentRecord?
            do {
                record = try settings.enrollment()
            } catch {
                guard revision == credentialRevision, !Task.isCancelled else { return }
                credentialEligible = false
                await client.clearConfigurationAndStop(credentialRevision: revision)
                guard revision == credentialRevision, !Task.isCancelled else { return }
                notice = settings.storageError
                return
            }
            if let record {
                await configureAndStart(record, credentialRevision: revision)
                return
            }

            guard allowAutomaticEnrollment, !automaticEnrollmentAttempted else {
                relaySetupRetryAvailable = true
                notice = "Relay setup needs attention. Retry setup to create a new private session."
                await client.clearConfigurationAndStop(credentialRevision: revision)
                return
            }
            automaticEnrollmentAttempted = true
            await performEnrollment(
                credentialRevision: revision,
                automaticRecoveryUsed: settings.hasPartialLegacyEnrollment
            )
        }
        credentialTask = task
        return task
    }

    private func enrollAndConnect(
        credentialRevision revision: UInt64,
        automaticRecoveryUsed: Bool
    ) -> Task<Void, Never> {
        let task = Task {
            await performEnrollment(
                credentialRevision: revision,
                automaticRecoveryUsed: automaticRecoveryUsed
            )
        }
        credentialTask = task
        return task
    }

    private func performEnrollment(
        credentialRevision revision: UInt64,
        automaticRecoveryUsed: Bool
    ) async {
        notice = automaticRecoveryUsed
            ? "Replacing the private relay session…"
            : "Creating your private relay session…"
        do {
            let enrollment = try await enroll()
            guard revision == credentialRevision, credentialEligible, !Task.isCancelled else { return }
            let record = try settings.saveEnrollment(
                enrollment,
                automaticRecoveryUsed: automaticRecoveryUsed
            )
            guard revision == credentialRevision, credentialEligible, !Task.isCancelled else { return }
            notice = automaticRecoveryUsed
                ? "Private relay session replaced."
                : "Private relay session created."
            relaySetupRetryAvailable = false
            await configureAndStart(record, credentialRevision: revision)
        } catch {
            guard revision == credentialRevision, !Task.isCancelled else { return }
            if settings.storageError != nil {
                credentialEligible = false
            }
            await client.clearConfigurationAndStop(credentialRevision: revision)
            guard revision == credentialRevision, !Task.isCancelled else { return }
            relaySetupRetryAvailable = true
            notice = settings.storageError ?? error.localizedDescription
        }
    }

    private func configureAndStart(
        _ record: DesktopEnrollmentRecord,
        credentialRevision revision: UInt64
    ) async {
        guard revision == credentialRevision, credentialEligible else { return }
        do {
            let relayURL = try RelayEndpoint.validated(record.relayURL, allowLocalSimulator: false)
            let pairing = PairingController(transport: client, relayURL: relayURL)
            self.pairing = pairing
            phoneManagement = PhoneManagementController(transport: client)
            pairingAction = PairingActionPresentation(status: nil)
            let configured = try await client.configure(
                urlString: record.relayURL,
                token: record.macToken,
                allowLocalSimulator: false,
                credentialRevision: revision
            )
            guard configured else { return }
            try Task.checkCancellation()
            guard revision == credentialRevision, credentialEligible else { return }
            await client.start(credentialRevision: revision)
        } catch {
            guard revision == credentialRevision, !Task.isCancelled else { return }
            notice = "Saved relay enrollment is invalid: \(error.localizedDescription)"
            relaySetupRetryAvailable = true
        }
    }

    private func recoverAfterAuthenticationRejection() {
        let record: DesktopEnrollmentRecord
        do {
            guard let stored = try settings.enrollment() else {
                relaySetupRetryAvailable = true
                notice = "Relay authentication failed. Retry setup to create a new private session."
                return
            }
            record = stored
        } catch {
            credentialEligible = false
            relaySetupRetryAvailable = false
            notice = settings.storageError
            return
        }

        guard !record.automaticRecoveryUsed else {
            relaySetupRetryAvailable = true
            notice = "Automatic relay recovery was rejected. Retry setup when you are ready."
            return
        }

        let revision = beginCredentialOperation()
        relaySetupRetryAvailable = false
        _ = enrollAndConnect(credentialRevision: revision, automaticRecoveryUsed: true)
    }
}

File: windows/src/ClickBridge.Windows/AppState.cs
using ClickBridge.Core.Actions;
using ClickBridge.Core.Pairing;
using ClickBridge.Core.Relay;
using ClickBridge.Core.Storage;
using ClickBridge.Core.Wire;

namespace ClickBridge.Windows;

/// <summary>
/// Port of mac/ClickBridgeMac/AppState.swift's glue logic. Swift's version
/// crosses several actor boundaries (RelayClient, ActionProcessor) so many
/// of its steps are individually awaited; here those callees are plain
/// synchronous methods, so the constructor wiring collapses to direct calls
/// and only the genuinely-async credential/connect flow stays as a
/// fire-and-forget Task, matching the same revision-fencing behavior.
/// Raises Changed after any property update so the UI layer (TrayApplicationContext)
/// can refresh — callers must marshal that back to the UI thread themselves.
/// </summary>
public sealed class AppState
{
    public RelayClient.Status Connection { get; private set; } = RelayClient.Status.Disconnected;
    public PermissionState Permission { get; private set; }
    public string LastResult { get; private set; } = "—";
    public string? Notice { get; private set; }
    public PairingController? Pairing { get; private set; }
    public PairingActionPresentation PairingAction { get; private set; } = PairingActionPresentation.From(null);

    public SettingsStore Settings { get; }

    public event Action? Changed;

    private readonly RelayClient _client;
    private readonly ActionProcessor _processor;
    private readonly IPostEventPermissionChecking _permissionService;

    private ulong _credentialRevision;
    private bool _credentialEligible = true;
    private ulong _lastStatusSequence;

    public AppState(
        SettingsStore settings,
        RelayClient client,
        ActionProcessor processor,
        IPostEventPermissionChecking permissionService,
        string? initialNotice = null)
    {
        Settings = settings;
        _client = client;
        _processor = processor;
        _permissionService = permissionService;
        Notice = initialNotice;
        Permission = permissionService.IsGranted() ? PermissionState.Ready : PermissionState.Required;

        client.SetStatusHandler(HandleStatusEvent);
        client.SetResultHandler(HandleResult);
        client.SetPairingHandler(ReceivePairingAsync);
        processor.SetRemoteEnabled(settings.RemoteEnabled);

        var initialCredentialRevision = BeginCredentialOperation();
        _ = InitializeAsync(initialCredentialRevision);
    }

    private async Task InitializeAsync(ulong initialCredentialRevision)
    {
        await PublishStateAsync().ConfigureAwait(false);
        if (initialCredentialRevision != _credentialRevision) return;
        await ConnectAsync(initialCredentialRevision).ConfigureAwait(false);
    }

    public Task ReconnectAsync() => ConnectAsync(BeginCredentialOperation());

    public Task PublishStateAsync() =>
        _client.UpdateAdvertisedStateAsync(new MacState { RemoteEnabled = Settings.RemoteEnabled, Permission = Permission });

    public void SetPersistedRemoteEnabled(bool enabled)
    {
        _ = Task.Run(async () =>
        {
            _processor.SetRemoteEnabled(enabled);
            Settings.RemoteEnabled = enabled;
            await PublishStateAsync().ConfigureAwait(false);
            RaiseChanged();
        });
    }

    public void RefreshPermission()
    {
        // Unlike macOS Accessibility, Windows has no permission that can be
        // revoked behind the app's back, so there is no "app became active"
        // hook to re-check here — WindowsInputPermissionService always
        // reports granted; a real SendInput failure surfaces per-action
        // instead (see WindowsInputExecutor).
        Permission = _permissionService.IsGranted() ? PermissionState.Ready : PermissionState.Required;
        _ = PublishStateAsync();
        RaiseChanged();
    }

    public void RequestPermission() => RefreshPermission();

    public void BeginPairing()
    {
        if (Pairing is { } pairing)
        {
            _ = pairing.BeginPairingAsync();
        }
    }

    public void ConfirmReplacement()
    {
        if (Pairing is not { } pairing) return;
        _ = Task.Run(async () =>
        {
            await pairing.BeginPairingAsync().ConfigureAwait(false);
            await pairing.ConfirmReplacementAsync().ConfigureAwait(false);
        });
    }

    public Task SaveTokenAsync(string token)
    {
        var revision = BeginCredentialOperation();
        try
        {
            Settings.SaveMacToken(token);
            _credentialEligible = true;
            Notice = "Token saved.";
            RaiseChanged();
            return ConnectAsync(revision);
        }
        catch
        {
            _credentialEligible = false;
            Notice = Settings.StorageError;
            RaiseChanged();
            return _client.ClearConfigurationAndStopAsync(revision);
        }
    }

    public async Task ClearTokenAsync()
    {
        var revision = BeginCredentialOperation();
        _credentialEligible = false;
        string? persistenceError = null;
        try
        {
            Settings.ClearMacToken();
        }
        catch
        {
            persistenceError = Settings.StorageError;
        }

        if (revision != _credentialRevision) return;
        var cleared = await _client.ClearConfigurationAndStopAsync(revision).ConfigureAwait(false);
        if (!cleared || revision != _credentialRevision) return;
        Notice = persistenceError ?? "Token cleared.";
        RaiseChanged();
    }

    private void HandleStatusEvent(RelayClient.StatusEvent evt)
    {
        if (evt.CredentialRevision != _credentialRevision || evt.Sequence <= _lastStatusSequence) return;
        _lastStatusSequence = evt.Sequence;
        Connection = evt.Status;
        RaiseChanged();

        if (Pairing is not { } pairing) return;
        var capabilityAvailable = evt.Status == RelayClient.Status.Connected;
        _ = Task.Run(async () =>
        {
            await pairing.RefreshStatusAsync(capabilityAvailable).ConfigureAwait(false);
            if (evt.CredentialRevision != _credentialRevision || evt.Sequence != _lastStatusSequence) return;
            RaiseChanged();
        });
    }

    private void HandleResult(ActionResult result)
    {
        LastResult = $"{result.Status.ToWire()}: {result.Reason.ToWire()}";
        RefreshPermission();
    }

    private async Task ReceivePairingAsync(WireMessage message)
    {
        if (Pairing is not { } pairing) return;
        var previousState = pairing.State;
        await pairing.ReceiveAsync(message).ConfigureAwait(false);
        if (message is PairStatus status && previousState is PairingState.CheckingStatus && pairing.State is PairingState.Ready)
        {
            PairingAction = PairingActionPresentation.From(status);
            RaiseChanged();
        }
    }

    private async Task ConnectAsync(ulong revision)
    {
        if (revision != _credentialRevision) return;
        if (!_credentialEligible)
        {
            await _client.ClearConfigurationAndStopAsync(revision).ConfigureAwait(false);
            return;
        }

        string token;
        try
        {
            var storedToken = Settings.MacToken();
            if (string.IsNullOrEmpty(storedToken))
            {
                if (revision != _credentialRevision || !_credentialEligible) return;
                Notice = "Save MAC_TOKEN in Settings before connecting.";
                RaiseChanged();
                return;
            }
            token = storedToken;
        }
        catch
        {
            if (revision != _credentialRevision) return;
            _credentialEligible = false;
            await _client.ClearConfigurationAndStopAsync(revision).ConfigureAwait(false);
            if (revision != _credentialRevision) return;
            Notice = Settings.StorageError;
            RaiseChanged();
            return;
        }

        if (revision != _credentialRevision || !_credentialEligible) return;
        try
        {
            var relayUrl = RelayEndpoint.Validated(Settings.RelayUrlString, allowLocalSimulator: false);
            var pairing = new PairingController(_client, relayUrl);
            Pairing = pairing;
            PairingAction = PairingActionPresentation.From(null);
            var configured = await _client.ConfigureAsync(Settings.RelayUrlString, token, allowLocalSimulator: false, revision)
                .ConfigureAwait(false);
            if (!configured) return;
            if (revision != _credentialRevision || !_credentialEligible) return;
            _client.Start(revision);
            RaiseChanged();
        }
        catch (Exception ex)
        {
            if (revision != _credentialRevision) return;
            Notice = $"Connection settings are invalid: {ex.Message}";
            RaiseChanged();
        }
    }

    private ulong BeginCredentialOperation()
    {
        _credentialRevision += 1;
        return _credentialRevision;
    }

    private void RaiseChanged() => Changed?.Invoke();
}
